### ISSUE TITLE
🚀 Improve the SolidStart Cloudflare (bindings) integration

### ISSUE THREAD

**Issue by dario-piotrowicz** [cloudflare/workers-sdk#5912]
## Summary

People deploying their SolidStart applications to Cloudflare Pages do not have a proper DX when it comes to accessing [Bindings](https://developers.cloudflare.com/pages/functions/bindings/), it would be great if we could find a solution for this.

> [!NOTE]
> - SolidStart is currently the only mainstream fullstack framework that doesn't provide a good Cloudflare bindings DX
>    (all mainstream fullstack frameworks reply on the [`getPlatformProxy`](https://developers.cloudflare.com/workers/wrangler/api/#getplatformproxy) utilities in order to provide bindings locally)
> - SolidStart prior to the Beta 2 reimplementation had a very good Cloudflare bindings DX 
> - SolidStart applications _not_ using bindings work perfectly fine on Cloudflare Pages

## Details

Prior to the [SolidStart Beta 2](https://github.com/solidjs/solid-start/discussions/1052) reimplementation users were able to use bindings via an `env` object made available by utilities such as `createServerData$`, for example:
```ts
export function routeData() {
  return createServerData$(async (_, { env }) => {
    const { MY_KV } = (env as { MY_KV: KVNamespace }));
    // ...
  });
}
```

Such bindings were available both during local development and in production.

For local development people would declare their bindings in the vite.config file, for example:
```ts
import solid from "solid-start/vite";
import { defineConfig } from "vite";
import cloudflare from "solid-start-cloudflare-pages";

export default defineConfig({
  plugins: [
   solid({
     adapter: cloudflare({
       kvNamespaces: ["MY_KV"],
     }),
   }),
  ],
});
```

and the vite plugin would, by using Miniflare (2) populate such bindings during local development.

Developers could then also configure their bindings in the Cloudflare dashboard and have them available in production as well.

Since SolidStart Beta 2 however the above API is no longer available and:
 - there is no way to access Bindings locally
 - the way of accessing bindings looks like a very poor/hacky DX
 
In details:

### There is no way to access Bindings locally

SolidStart, since Beta 2 is using [vinxi](https://github.com/nksaraf/vinxi) under the hood, vinxi is (as far as my understanding goes) a fork of [nitro](https://nitro.unjs.io/), although in nitro we do have a solution ([nitro-cloudflare-dev](https://github.com/pi0/nitro-cloudflare-dev)) to make Cloudflare bindings available during local development, such solution doesn't work in vinxi because of how it diverged from nitro.

### The way of accessing bindings looks like a very poor/hacky DX

In the current SolidStart implementation, in server side code the Cloudflare bindings are accessible via the [`nativeEvent`](https://docs.solidjs.com/solid-start/advanced/request-events#nativeevent) object in the following way:
```ts
apiEvent.nativeEvent.context.cloudflare.env.MY_KV
```
This feels very unpolished and ideally people wouldn't have to use such an API to access Cloudflare bindings.

Also personally I am not sure if this `nativeEvent` is accessible throughout all the application's server side code (it might as well be, but I haven't really investigated this deeply).

## What to do

We should try to collaborate with the SolidStart team and:
1. find a way for SolidStart application to be able to use bindings during local development, by either re-aligning vinxi with nitro and make use of `nitro-cloudflare-dev`, or by having vinxi/SolidStart use `getPlatformProxy` directly (which is the utility used under the hood by `nitro-cloudflare-dev`)
2. trying to implement a more proper API for accessing Cloudflare bindings that doesn't necessarily require to go through the `nativeEvent`, if we can't manage to agree on/implement a better API, the current one could be good enough, but ideally we'd have to make sure that `nativeEvent` is accessible throughout all the application's server-side code
3. update our [C3 solid scaffolding](https://github.com/cloudflare/workers-sdk/blob/5bf0a6b7fa365305143f0d1ce1426bb55bc3a085/packages/create-cloudflare/templates/solid) based on the outcome of the above
4. update out [solid-start guide](https://developers.cloudflare.com/pages/framework-guides/deploy-a-solid-site/) based on the outcome of the above

____

> [!NOTE]
> This issue is for tracking purposes and involves more than the `workers-sdk` repository (but this was what seemed the most convenient place where to put this)

___

## Related issues

Related: https://github.com/cloudflare/cloudflare-docs/issues/14222
Related: https://github.com/cloudflare/cloudflare-docs/issues/14679

**Comment by dario-piotrowicz** [comment]
@Tommypop2 if you could help out, especially with points `1.` and `2.` that would be really great 🙏

I am more than happy to get involved and be part of any conversation regarding potential solutions and decisions 🙂

@pi0 (the main nitro maintainer) has also had conversations with the SolidStart team in the past and might provide great insights regarding the vinxi situation and potentially get involved in conversations/decisions regarding that 🙂

**Comment by IgorMinar** [comment]
@ryansolid fyi - per our convo we are looking into updating the integration. @Tommypop2 graciously offered to take the lead on this one with @dario-piotrowicz supporting him. 🚀

**Comment by pi0** [comment]
Vinxi dev server is not same as nitro's isolated dev-server and that's what makes this inconsistencies.

Nitro v3 will adopt Vite and also built-in Worked isolation based on same model of Node worker isolation. So ideal and stable solution would be that Solid-start/vinxi adopts this and avoids non-isolated dev runtime.

**Comment by jcramb** [comment]
Hi all :) Happy to see this being acknowledged as an issue - any progress on plans to improve the DX for this?

**Comment by peterhirn** [comment]
I'm maintaining this demo app running SolidStart on Cloudflare Pages: https://github.com/phi-ag/solid-pages

It uses a middleware and `getPlatformProxy` to unify access to eg. bindings on `event.locals`, see https://github.com/phi-ag/solid-pages/blob/main/src/middleware.ts#L47

Accessing `env` globally doesn't work yet:
- https://developers.cloudflare.com/changelog/2025-03-17-importable-env/
- https://github.com/phi-ag/solid-pages/pull/335

Going forward `getPlatformProxy` is hopefully going to be replaced with running everything natively in `workerd` using  Vite 6 Environment API, see https://github.com/solidjs/solid-start/discussions/1743

**Comment by jcramb** [comment]
Thanks for the link to the blog post, I don't know how I missed it! Being able to import `env` is a fantastic improvement for my current use case - initializing a better auth client which can be accessed during requests.

**Comment by tuzemec** [comment]
Ok, so as far as I get it, if I create a new SolidStart Cloudflare project as described [here](https://developers.cloudflare.com/workers/frameworks/framework-guides/solid/), there's no easy way to access the bindings and thus the D1 database configured in my wrangler config?

Appologies, just trying to wrap my head around all this :)

**Comment by jcramb** [comment]
It's easy enough to access the bindings during a request i.e. in a middleware or anywhere you can call solid's `getRequestEvent()` via `event.nativeEvent.context.cloudflare.env` which will be of type `Cloudflare.Env` as defined by the worker config generated by `wrangler types`.

It's tricky to access the bindings globally, outside of a request, which is problematic when you need to initialize certain things like auth handlers etc. In my use case I'm integrating BetterAuth but thankfully cloudflare now exports the bindings via `import { env } from "cloudflare:workers";` so that helps my use case a bit - just don't try to actually interact with the db until you're in a request.

**Comment by tuzemec** [comment]
@jcramb thanks a lot!

I'm trying to do exactly that - add BetterAuth. And I'm getting errors that 'cloudflare:workers' can't be resolved when I try to use the `@better-auth/cli`, but that's to be expected?

**Comment by jcramb** [comment]
> [@jcramb](https://github.com/jcramb) thanks a lot!
> 
> I'm trying to do exactly that - add BetterAuth. And I'm getting errors that 'cloudflare:workers' can't be resolved when I try to use the `@better-auth/cli`, but that's to be expected?

Yeah I haven't found a great way to use @better-auth/cli with cloudflare yet - it works great with the library itself though so i've just defined the schema / migrations myself using their documentation.

**Comment by petebacondarwin** [comment]
Closing this old issue that has not seen any activity for over a year.
